Add support for `preferred_element_type` arg in convolutions

lgeiger opened the pull request:
XLA recently added support for integer convolutions on CPU in https://github.com/tensorflow/tensorflow/commit/89fc1ddd96f25aa9d8eb6a0104ca1fdb11b26753.

This PR adds support for the `preferred_element_type` parameter in the `conv`, `conv_general_dilated`, `conv_with_general_padding`, `conv_general_dilated_patches` and `conv_transpose` which is passed through to XLA.

This PR is analogous to #5213 which added support for `preferred_element_type` in `lax.dot`.

Please note that I don't know if this has any implications for GPU since GPUs don't have support integer convolution yet, but hopefully CI will catch any potential problems related to GPUs. Please let me know if I am missing something.

Do you know if there is a timeline integer GPU convolution support in XLA which would be very useful for research on quantized neural networks?

jakevdp commented on jax/_src/lax/lax.py:
```diff
@@ -2810,11 +2830,25 @@ def _conv_general_dilated_shape_rule(
 
+def _validate_preferred_elemet_type(input_dtype, preferred_element_type):
+  if dtypes.issubdtype(input_dtype, np.integer) and not dtypes.issubdtype(preferred_element_type, np.integer):
+    raise TypeError("`preferred_element_type` and the original type must both be integral or both be floating point.")
```
It looks like this doesn't catch the case when `input_dtype` is a float and `preferred_element_type` is an integer.

Do we need a similar check for complex?

lgeiger replied:
> It looks like this doesn't catch the case when input_dtype is a float and preferred_element_type is an integer.

Good catch! This checking was just copied from `lax.dot` The interpreter throws a sensible error message, but I agree it's better to catch it already here.

jakevdp reviewed (commented):
Looks great, thanks! Just one minor comment.

lgeiger commented:
Thanks for the fast review 🚀 

I improved the `dtype` checking in 2ec096d08f5220fc299ba807b767f36777a4a66e and extended it to also handle complex types.

lgeiger commented:
@jakevdp Out of curiosity, do you know whether there is support for integer convolutions on GPUs planned in XLA? I think this could be incredibly useful for research on quantized neural networks if XLA is able generate fast integer kernels.

jakevdp commented:
> do you know whether there is support for integer convolutions on GPUs planned in XLA?

I'm not certain... I see you're already following #4012 which is probably the relevant issue for that.

gnecula commented:
@lgeiger I have not seen integer convolutions in the plans for GPU for the next quarter. Have you filed a feature request with XLA?

lgeiger commented:
I wasn't aware of the complex number support so 7870ecb0c48f9d2753f322270b56c811cdfd4bf1 fixes the handling of complex numbers when `expand_complex_convolutions=True` and extends the unittests.

> I have not seen integer convolutions in the plans for GPU for the next quarter. Have you filed a feature request with XLA?

Thanks for the info. No, I haven't filled an issue yet. Would `tensorflow/tensorflow` be the right place for this?
I've only seen a reference to an internal issue so far:
https://github.com/google/jax/blob/0f1ddf6dc39ea29d39b256d4009406c6c1c5c844/tests/lax_test.py#L496

lgeiger commented on jax/_src/lax/lax.py:
```diff
@@ -2940,18 +2975,38 @@ def _conv_general_dilated_translation_rule(
+    if preferred_element_type is not None:
+      # Convert complex dtype to types used for real and imaginary parts
+      assert np.issubdtype(preferred_element_type, np.complexfloating)
+      preferred_element_type = np.dtype(f"f{np.dtype(preferred_element_type).itemsize // 2}")
```
Not sure if there is a nicer way of doing this. I'm very open to suggestions.

lgeiger commented on jax/_src/lax/lax.py:
```diff
@@ -2940,18 +2975,38 @@ def _conv_general_dilated_translation_rule(
+      # Convert complex dtype to types used for real and imaginary parts
+      assert np.issubdtype(preferred_element_type, np.complexfloating)
+      preferred_element_type = xla_client.dtype_to_etype(
+          np.dtype(f"f{np.dtype(preferred_element_type).itemsize // 2}"))
```
Not sure if there is a nicer way of doing this. I'm very open to suggestions :)

jakevdp replied:
I'd use a helper function, similar to this: https://github.com/google/jax/blob/dbe9ff5f2657f9c124f60e4183d05423d913af8a/jaxlib/rocsolver.py#L36-L47

jakevdp replied:
or, since you know this is complex and there are only two possibilities, an if/else statement would probably suffice :grin:

lgeiger replied:
Great idea! I wasn't aware that there are only two complex types. Fixed in c8f4173b6ec3093d8a3130a26fe31a0e764dc963

jakevdp reviewed (approved):
Looks great - thanks!

Final request: can you squash all the changes into a single commit and force-push to the branch? (see https://github.com/google/jax/blob/master/CONTRIBUTING.md)

lgeiger commented:
> Final request: can you squash all the changes into a single commit and force-push to the branch?

Ah sorry missed that, I thought that would happen during merge via GitHubs "Squash and merge" option. I rebased the PR and squashed everything into one commit.

jakevdp commented:
It looks like this is causing some failures in `jax2tf`, because it doesn't handle the new `preferred_element_type` argument (these tests are not currently run in by the github CI).

You can run them locally by making sure you have the most recent version of tensorflow installed, and run `pytest -n auto jax/experimental/jax2tf/tests`

lgeiger commented:
> It looks like this is causing some failures in jax2tf, because it doesn't handle the new preferred_element_type argument (these tests are not currently run in by the github CI).

:+1: I updated `jax2tf` and rebased the PR.  Let's see what CI thinks

jakevdp commented on tests/lax_test.py:
```diff
@@ -62,6 +62,18 @@
+  (np.float64, np.float64), (np.int8, np.int8), (np.int8, np.int16), (np.int8, np.int32),
+  (np.int8, np.int64), (np.int16, np.int16), (np.int16, np.int32), (np.int16, np.int64),
+  (np.int32, np.int32), (np.int32, np.int64), (np.int64, np.int64),
+  (np.complex64, np.complex64), (np.complex64, np.complex128), (np.complex128, np.complex128)]
```
`np.complex128` is currently not supported on TPU, so we need to avoid this test case for that device (you can check via `jtu.device_under_test() == 'tpu'`)

jakevdp replied:
Also on CPU we're seeing some failures due to insufficient precision:
```
LaxTest.testConvPreferredElement_lhs_shape=complex64[2,2,9,10]_rhs_shape=complex64[3,2,4,5]_preferred_element_type=<class 'numpy.complex128'>
```
You can probably reproduce this locally with
```
$ JAX_NUM_GENERATED_CASES=50 pytest -n auto tests/lax_test.py -k PreferredElement
```
Also worth testing with X64 enabled:
```
$ JAX_ENABLE_X64=1 JAX_NUM_GENERATED_CASES=50 pytest -n auto tests/lax_test.py -k PreferredElement
```

lgeiger replied:
Thanks for checking internal CI. I skipped the test on TPU and reduced the precision on CPU for the mentioned cases which should fix the failures.

jakevdp reviewed (commented):
Thanks! Just a couple more test failures to address on the more comprehensive non-github CI run.

lgeiger committed:
Add support for `preferred_element_type` arg in convolutions

jakevdp commented on tests/lax_test.py:
```diff
@@ -378,6 +390,53 @@ def fun(lhs, rhs):
+       (dtype == np.complex128 or preferred_element_type == np.complex128)):
+      raise SkipTest("np.complex128 is not yet supported on TPU")
+    tol = None
+    if (jtu.device_under_test() == "cpu" and
```
I'm actually seeing the same issue on GPU as well (sorry I didn't notice that earlier) so we may as well set this tolerance for all devices

jakevdp commented on tests/lax_test.py:
```diff
@@ -921,16 +980,7 @@ def testDot(self, lhs_shape, rhs_shape, dtype, precision):
-        (np.int8, np.int64), (np.int16, np.int16), (np.int16, np.int32), (np.int16, np.int64),
-        (np.int32, np.int32), (np.int32, np.int64), (np.int64, np.int64)]))
+      for dtype, preferred_element_type in preferred_type_combinations))
   def testDotPreferredElement(self, lhs_shape, rhs_shape, dtype, preferred_element_type):
```
complex128 on TPU needs to be skipped here as well

jakevdp reviewed (commented):
So close! Just a couple more CI issues.

lgeiger commented:
Thanks for the help. I updated the tests to handle the complex cases on GPU and TPU.

jakevdp commented:
Thanks for sticking with this! Looks like we can get this merged now.

lgeiger commented:
> Thanks for sticking with this! Looks like we can get this merged now.

Sure, thanks for the help and the fast review. Really appreciate it 👍

lgeiger commented:
> I have not seen integer convolutions in the plans for GPU for the next quarter. Have you filed a feature request with XLA?

For reference, I now opened an issue at https://github.com/tensorflow/tensorflow/issues/49140
